zkfold-cardano-scripts/src/ZkFold/Cardano/UPLC/Wallet.hs

```diff
@@ -1,103 +1,85 @@
-wallet zkpCheck WalletSetup{..} WalletRedeemer{..} ctx@(ScriptContext TxInfo{..} _ scriptInfo) =
-    case (scriptInfo, maybeScriptHash) of
-      (SpendingScript _ _, Just scriptHash) -> forwardingReward scriptHash () ctx
-      (SpendingScript _ _, _)               -> False
-      _                                     -> case wrCreds of
-            SpendWithSignature sign -> any (== sign) $ getPubKeyHash <$> txInfoSignatories -- pubKayHash is present in the signatories list
-            SpendWithWeb2Token w2c  -> zkpPasses w2c && outputsCorrect w2c
-    where
-        maybeScriptHash = do
-            inp <- findOwnInput ctx
-            let addrCreds = addressCredential . txOutAddress . txInInfoResolved $ inp
-            case addrCreds of
-              ScriptCredential scr -> pure $ getScriptHash scr
-              _                    -> Nothing
-
-        compressedPI Web2Creds{..} = toInput . blake2b_224 $ foldl1 appendByteString [wUserId, wTokenHash, fromInput . toF $ wAmount, wrTxRecipient]
-        zkpPasses w2c = verify @PlonkupPlutus @HaskellCore zkpCheck (compressedPI w2c) wrZkp
-
-        outputsCorrect Web2Creds {..} = and
-            [ length txInfoOutputs == 2                           -- only two outputs
-            , all ((== NoOutputDatum) . txOutDatum) txInfoOutputs -- datums are empty
-            , lovelaceValue (Lovelace wAmount) == txOutValue (head txInfoOutputs)  -- the amount is correct
-            , wrTxRecipient == (getCredential . txOutAddress) (head txInfoOutputs) -- the recipient's address is correct
-            , wsPubKeyHash == (getCredential . txOutAddress) (txInfoOutputs !! 1)  -- the wallet's change address is correct
-            ]
-
-        getCredential :: Address -> BuiltinByteString
-        getCredential (Address (PubKeyCredential pkey) _) = getPubKeyHash pkey
-        getCredential (Address (ScriptCredential scr)  _) = getScriptHash scr
-
-
-{-# INLINABLE untypedWallet #-}
-untypedWallet :: SetupBytes -> WalletSetup -> BuiltinData -> BuiltinUnit
-untypedWallet zkpCheck setup ctx' =
+import           ZkFold.Cardano.UPLC.Wallet.Types
+
+-- TODO: Account for rotation of public keys
+-- TODO: Check the client Id
+-- TODO: Check the suffix length (must be a predefined size)
+-- TODO: Do we need to split bytestrings further due to ledger rules?
+{-# INLINEABLE web2Auth #-}
+-- | Mints tokens paramterized by the user's email and a public key selected by the user.
+web2Auth :: SetupBytes -> Web2Creds -> ScriptContext -> Bool
+web2Auth expModCircuit Web2Creds {..} (ScriptContext TxInfo {..} red (MintingScript symb)) =
+  let
+    publicInput = toInput (sha2_256 $ jwtPrefix <> w2cEmail <> jwtSuffix) * toInput bs
+    Web2Auth JWTParts {..} proof tn@(TokenName bs) = unsafeFromBuiltinData . getRedeemer $ red
+   in
+    -- Check that the user knows an RSA signature for a JWT containing the email
+    verify @PlonkupPlutus @HaskellCore expModCircuit publicInput proof
```

Comment:
How should `expModCircuit` work? Should it receive a single field element and raise it to the power of public exponent?